drivers: console: posix_arch_console: build warning with `CONFIG_PRINTK=n`

### Describe the bug

When building for `native_sim` with `CONFIG_PRINTK=n`, a build warning is emitted from compiling `drivers/console/posix_arch_console.c`.

### Regression

- [ ] This is a regression.

### Steps to reproduce

Steps to reproduce:
1) `west build -b native_sim/native/64 samples/basic/blinky -- -DCONFIG_PRINTK=n -DCONFIG_BOOT_BANNER=n`
2) See build error below

### Relevant log output

```shell
[ 90%] Building C object zephyr/drivers/console/CMakeFiles/drivers__console.dir/posix_arch_console.c.obj
/home/brix/Projects/zephyrproject/zephyr/drivers/console/posix_arch_console.c:17:12: warning: ‘print_char’ defined but not used [-Wunused-function]
   17 | static int print_char(int c)
      |            ^~~~~~~~~~
```

### Impact

Annoyance – Minor irritation; no significant impact on usability or functionality.

### Environment

- OS: Linux
- SDK: Zephyr SDK 0.17.2
- Zephyr SHA: a1f66f05d292ff51c3c6fdc081867bc0a5b232bd

### Additional Context

Disabling `CONFIG_PRINTK` is recommended for hardened release configs. This bug prevents building these hardened configs for native_sim in downstream CI without warnings.

--- a/drivers/console/posix_arch_console.c
+++ b/drivers/console/posix_arch_console.c
@@ -14,6 +14,7 @@
 static char stdout_buff[_STDOUT_BUF_SIZE];
 static int n_pend; /* Number of pending characters in buffer */
 
+#if defined(CONFIG_PRINTK) || defined(CONFIG_STDOUT_CONSOLE)
 static int print_char(int c)
 {
 	int printnow = 0;
@@ -36,6 +37,7 @@ static int print_char(int c)
 	}
 	return c;
 }
+#endif /* defined(CONFIG_PRINTK) || defined(CONFIG_STDOUT_CONSOLE) */
 
 /**
  * Ensure that whatever was written thru printk is displayed now

